Script to ghstack land

kirklandsign opened the pull request:
Per ghstack [tutorial](https://github.com/ezyang/ghstack/blob/master/README.md#structure-of-submitted-pull-requests) for a PR exported by ghstack, it will have two diff sets:
- `gh/user/1/base <- gh/user/1/head` where the PR is created like [this](https://github.com/pytorch/executorch/pull/6265) 
- `main <- gh/user/1/orig`

The purpose of this bot is, when the ghstack PR is merged, automatically create another PR to do this merge `main <- gh/user/1/orig`. Then we just merge the newly created PR so that `main` has that change.

Missing piece: token from pytorch bot to create PR.

If this goes well, we can either `git merge` the commit from gh/user/1/orig into main directly, without going through the new PR; or auto approve and merge the PR.

Test: You can test locally with `export GITHUB_TOKEN=ghpxyz; python .github/scripts/propose_ghstack_orig_pr.py --pr 6265 --repo pytorch/executorch`
Note that between /orig merges, there is never a merge conflict.

The invariant is guaranteed that gh/user/1/orig contains the exact same change between `gh/user/1/base <- gh/user/1/head` by ghstack tool.

kirklandsign commented on .github/workflows/ghstack_land.yml:
```diff
@@ -0,0 +1,30 @@
+
+        python .github/scripts/propose_ghstack_orig_pr.py --pr $PR_NUMBER --repo pytorch/executorch
+      env:
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
@huydhn could use your help with getting a bot token for creating the PR

kirklandsign replied:
@huydhn Then let's still use GH_PYTORCHBOT_CHERRY_PICK_TOKEN, and add environment: cherry-pick-bot?

kirklandsign commented on .github/workflows/ghstack_land.yml:
```diff
@@ -0,0 +1,30 @@
+name: Propose to merge ghstack orig PRs to main
+on:
+  pull_request:
+    types: [opened, synchronize, closed]
```
very cheap job.

huydhn commented:
I need to sit down to review this one, but let me ask @kit1980 to take a look too as granting pull request write permission to a `pull_request` target seems like a security hole.

kirklandsign commented:
> I need to sit down to review this one, but let me ask @kit1980 to take a look too as granting pull request write permission to a `pull_request` target seems like a security hole. And he knows that part well.

Thank you! It’s very similar to our cherry pick bot. However if I use cherry pick bot token it doesn’t allow me to create PR (maybe on this branch trigger). GH actions token can’t be used to create PR.

kirklandsign commented on .github/workflows/ghstack_land.yml:
```diff
@@ -0,0 +1,28 @@
+
+        python .github/scripts/propose_ghstack_orig_pr.py --pr $PR_NUMBER --repo pytorch/executorch
+      env:
+        GITHUB_TOKEN: ${{ secrets.GH_PYTORCHBOT_CHERRY_PICK_TOKEN }}
```
@huydhn @kit1980 Can we accept `gh/*/base` to trigger the cherry pick bot?

Users already needs permission to push to pytorch:gh/*/base, so this won't grant additional permission

huydhn commented on .github/workflows/ghstack_land.yml:
```diff
@@ -0,0 +1,28 @@
+  pull_request:
+    types: [opened, synchronize, closed]
+    branches:
+      - 'gh/*/[0-9]+/base'
```
I think if you want to choose a smaller group of people to test this out first, you can list them here explicitly, i.e.

```
    branches:
      - 'gh/kirklandsign/[0-9]+/base'
      - 'gh/guangy10/[0-9]+/base'
```

It seems like a good idea to try it out first before rolling out to everyone.

kirklandsign committed:
Script to ghstack land

ghstack-source-id: 441280b30010718a5440390f5de05133f872728f
Pull Request resolved: https://github.com/pytorch/executorch/pull/6265

dbort commented on .github/scripts/propose_ghstack_orig_pr.py:
```diff
@@ -0,0 +1,135 @@
+    parser.add_argument(
+        "--pr",
+        type=int,
+        help="Number of the PR in the stack to check and create corresponding PR",
```
Can you clarify this? Is this an index into the stack, or the PR number itself? And if it's an existing PR, why does it need to create another PR?

kirklandsign replied:
That's the trigger PR. Example: https://github.com/pytorch/executorch/pull/6346

The bot creates https://github.com/pytorch/executorch/pull/6347 for merging into `main`
